### PR TITLE
Add requests wrappers

### PR DESCRIPTION
**Pull request by duncanmmacleod**
This PR introduces `igwn_auth_utils.requests`, which provides an IGWN auth mixin class for `requests.Session` to simplify use of the auth discovery methods in that context.

This is heavily influenced by `requests-gracedb` so @lpsinger should be involved in any discussion on reducing the overlap between that project and this one to provide a single `requests` wrapper for X.509/scitokens clients.

### CODE REVIEW

**Comment by duncanmmacleod** [comment]
@lpsinger, are you happy for me to include this 'new' code in this package? The longterm plan would be to have all gracedb clients use this module for auth token discovery.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]: This function gives me the willies. Could at least log the exception?

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]: Does it make any sense for both the cert and the scitoken to be used in the same session?

I would suggest organizing this into a single `if/elif` block like I did so that it is easy to verify that exactly one authentication method is set up:

```python
        # FIXME: these should go into elif clauses below
        # (as in `elif default_cert := _find_x509_credentials():`)
        # in order to defer unnecessary I/O,  but this requires
        # the := operator, which requires Python 3.8.
        default_cert = _find_cert()
        default_basic_auth = _find_username_password(url)

        if force_noauth:
            pass
        elif cert is not None:
            self.cert = cert
        elif username is not None:
            self.auth = (username, password)
        elif default_cert is not None:
            self.cert = default_cert
        elif default_basic_auth is not None:
            self.auth = default_basic_auth
        elif fail_if_noauth:
            raise ValueError('No authentication credentials found.')
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
We also need to support pulling http basic auth credentials from the netrc file, because that is how external collaborators get API access to GraceDB.

**Comment by duncanmmacleod** [comment]
@lpsinger, I think I've implemented your suggestions by making the cred discovery a bit more complicated, but also implementing the same basic auth as you did, and restricting auth to a single source (token OR x509 OR basic, in that order).

File: docs/_static/css/igwnauthutils.css
Comment (lpsinger) [inline docs/_static/css/igwnauthutils.css:1]: How is this stylesheet related to the present PR?

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]: You can eliminate the instance variable and property `_has_igwn_auth` if the `_init_auth` method returns a boolean or raises an exception.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]: Consider declaring a new exception class so that client code can catch this error explicitly. Take a look at existing Requests exception classes to see if there is one that would make sense as the base class.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py:165]: In the docstring, document the order of precedence in which all of the authentication methods are checked (and, for each method, the order in which environment variables and files are checked).

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]: I suggest using my `safe-netrc` package instead, which I developed in response to an LSC computing security review of gwcelery. Also, it will check the `NETRC` environment variable for you automatically.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py:169]: Like my `SessionAuthMixin` class, this should take an optional username and password for basic auth.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]:
```suggestion
        return bool(
            self.auth or
            "Authorization" in self.headers or
            self.cert
        )
```

This is a micro-optimization to take advantage of short-circuit evaluation.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py:292]: Not `raise`?

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]: See my `SessionErrorMixin` class.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py]: Thank you. This acknowledgement means a lot to me! Please add a link.

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — CHANGES_REQUESTED** [review]
A couple comments:

- It's not part of this PR, but I noticed that the order in which the `find_credentials` method checks environment variables differs from `ligo-gracedb` and `requests-gracedb`: your code checks `X509_USER_PROXY` first, whereas the others check `X509_USER_CERT` and `X509_USER_KEY` first. Please switch the order for compatibility.
- The auth initialization code is very cleanly written and doesn't depend much on `requests` internals, up until it is time to populate the object's member variables. Consider refactoring so that you can reuse the same initialization code for `requests`, `urllib3`, etc.
- Note that for autonomous, long-running processes, automatic refreshing of expiring credentials is an essential feature that is missing here. This is a concern for the old X.509 robot certificate model, but I do not know whether or not it is a concern for Scitokens.

File: docs/_static/css/igwnauthutils.css
Comment (duncanmmacleod) [reply docs/_static/css/igwnauthutils.css:1]: This improves the documentation which has been augmented to include docs of the new requests interface. It can easily be moved to a new PR.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: Yes, I pondered that then deliberately chose not to use it to eliminate a dependency. However, if there were significant security concerns about the builtin netrc (given that it now has the default file permission check) I can reconsider.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py:292]: No, in line with the other `_find_xxx` methods it doesn't ever fail if credentials aren't found.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: Nice, I will add that.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: Is this the best reference: https://requests-gracedb.readthedocs.io/?

**Comment by duncanmmacleod** [comment]
> A couple comments:
> 
> * It's not part of this PR, but I noticed that the order in which the `find_credentials` method checks environment variables differs from `ligo-gracedb` and `requests-gracedb`: your code checks `X509_USER_PROXY` first, whereas the others check `X509_USER_CERT` and `X509_USER_KEY` first. Please switch the order for compatibility.

This order is deliberately chosen to reproduce what is done by the globus toolkit tools. What is the motivation for checking separate cert/key before proxy?

> * The auth initialization code is very cleanly written and doesn't depend much on `requests` internals, up until it is time to populate the object's member variables. Consider refactoring so that you can reuse the same initialization code for `requests`, `urllib3`, etc.

I would prefer to defer that to a future PR, or until someone actually asks for that.

> * Note that for autonomous, long-running processes, automatic refreshing of expiring credentials is an essential feature that is missing here. This is a concern for the old X.509 robot certificate model, but I do not know whether or not it is a concern for Scitokens.

I have deliberately not attempted to tackle the auto-refresh stuff. As above I'd like to leave that to a future PR, or until I understand what this would entail for scitokens.

**Comment by lpsinger** [comment]
> > A couple comments:
> > 
> > * It's not part of this PR, but I noticed that the order in which the `find_credentials` method checks environment variables differs from `ligo-gracedb` and `requests-gracedb`: your code checks `X509_USER_PROXY` first, whereas the others check `X509_USER_CERT` and `X509_USER_KEY` first. Please switch the order for compatibility.
> 
> This order is deliberately chosen to reproduce what is done by the globus toolkit tools. What is the motivation for checking separate cert/key before proxy?

Uh oh. This sounds like a bug in ligo-gracedb and requests-gracedb. Would you please point me to the docs or the source that shows this to be the case, and open issues for those two packages?

> > * The auth initialization code is very cleanly written and doesn't depend much on `requests` internals, up until it is time to populate the object's member variables. Consider refactoring so that you can reuse the same initialization code for `requests`, `urllib3`, etc.
> 
> I would prefer to defer that to a future PR, or until someone actually asks for that.

Fair.

> > * Note that for autonomous, long-running processes, automatic refreshing of expiring credentials is an essential feature that is missing here. This is a concern for the old X.509 robot certificate model, but I do not know whether or not it is a concern for Scitokens.
> 
> I have deliberately not attempted to tackle the auto-refresh stuff. As above I'd like to leave that to a future PR, or until I understand what this would entail for scitokens.

OK. In the mean time, perhaps I can refactor requests-gracedb to use this package for the auth, and its own code for the cert reloading.

**Comment by duncanmmacleod** [comment]
> Uh oh. This sounds like a bug in ligo-gracedb and requests-gracedb. Would you please point me to the docs or the source that shows this to be the case, and open issues for those two packages?

Well, actually, this is more complicated that I thought. Globus tools separate the concept of a 'cert' vs a 'proxy', so `grid-proxy-info` will default to `X509_USER_PROXY`, however, `grid-cert-info` looks for `X509_USER_CERT`. Which makes sense.

You argument laid out [here](https://git.ligo.org/lscsoft/gracedb-client/-/issues/28) also makes sense (and I have tripped over that issue myself when using robot creds), but I would argue that in that case it is 'user-error' to define all of the X509 variables at the same time.

A quick google search has led me to some existing examples of the `PROXY > CERT` precedence, see [here](https://linux.die.net/man/1/htproxyrenew) from `GridSite` (whatever that is), or [here](https://github.com/italiangrid/voms/blob/v2.0.16/src/sslutils/sslutils.c#L2224-L2309) from the `voms` package.

To be clear, I have no horse in this race, and so don't have an opinion on what _should_ be correct.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: Done in latest commit.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: Done in latest commit, see the new `IgwnAuthError` object.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py:165]: Done in latest commit.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: Done in latest commit.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: Done in latest commit.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py]: This URL is included in the `intersphinx` configuration, so the references in the docstrings are automatically translated to hyperlinks in the docs.

**Comment by lpsinger** [comment]
> You argument laid out [here](https://git.ligo.org/lscsoft/gracedb-client/-/issues/28) also makes sense (and I have tripped over that issue myself when using robot creds), but I would argue that in that case it is 'user-error' to define all of the X509 variables at the same time.

Yet this caused problems with GWCelery in the past. Would you please conform to the order of precedence used by ligo-gracedb and requests-gracedb?

File: igwn_auth_utils/requests.py
Comment (lpsinger) [reply igwn_auth_utils/requests.py]: OK, would you please add an intersphinx ref in the docstring here?

File: docs/_static/css/igwnauthutils.css
Comment (lpsinger) [reply docs/_static/css/igwnauthutils.css:1]: Ok, that makes sense. I think it's fine to leave it in the same PR.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py:37]: Also document order in which environment variables, /tmp, and .globus dirs are checked.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py:354]: Don't modify kwargs. Or if you do, make a copy of it. The calling code probably does not expect this function to have side effects of modifying kwargs.

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by duncanmmacleod** [comment]
> Yet this caused problems with GWCelery in the past. Would you please conform to the order of precedence used by ligo-gracedb and requests-gracedb?

See https://github.com/duncanmmacleod/igwn-auth-utils/pull/11

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py:37]: This is referenced below:

```
    6.  look for an X.509 credential using
        :func:`igwn_auth_utils.find_x509_credential`
```

the sphinx docs will include a link to the documentation of the relevant function, which lists the ordering. I'm hesitant to duplicate that documentation here for fear that the two copies will get out of sync.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py:354]: `kwargs` is internal to the scope of this function and so cannot impact the calling code.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [reply igwn_auth_utils/requests.py:169]: This still doesn't have a username and password option.

File: igwn_auth_utils/requests.py
Comment (lpsinger) [inline igwn_auth_utils/requests.py:218]: What about falsey values like 0?

File: igwn_auth_utils/requests.py
Comment (lpsinger) [reply igwn_auth_utils/requests.py:354]: Oh, you're right! I learn something new every day...

```python
foo = {'bar': 1, 'bat': 2}

def xyzzy(**kwargs):
    kwargs.pop('bat')

xyzzy(**foo)
print(foo)  # outputs {'bar': 1, 'bat': 2}
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py:218]: I can only be bothered to protect the user from themselves so much. In this context I don't think confusing 0 for false is a problem, given that this isn't a valid token value, but the point is well taken.

File: igwn_auth_utils/requests.py
Comment (duncanmmacleod) [reply igwn_auth_utils/requests.py:169]: The `auth` option is a pre-existing space for a `(username, password)` tuple provided by the parent `requests.Session` object. The docstring for this code (now) clearly states that `auth`  is the correct keyword to use to pass in username/password. I would rather not overload the API with new keywords that provide the same functionality as existing ones.

**Review by lpsinger — APPROVED** [review]

**Pull request merged**
